src/nrnmpi/nrnmpi_dynam.cpp

```diff
@@ -33,221 +36,222 @@ extern "C" {
-        }();
-        // `handle` refers to "libmpi.so", figure out which MPI implementation that
-        // is.
-        auto const mpi_implementation = [handle] {
-            if (dlsym(handle, "ompi_mpi_init")) {
-                // OpenMPI
-                return "ompi";
-            } else if (dlsym(handle, "MPI_SGI_vtune_is_running")) {
-                // Got sgi-mpt. MPI_SGI_init exists in both mpt and hmpt, so we look
-                // for MPI_SGI_vtune_is_running which only exists in the non-hmpt
-                // version.
-                return "mpt";
-            } else {
-                // Assume mpich. Could check for MPID_nem_mpich_init...
-                return "mpich";
-            }
-        }();
-        auto const nrn_mpi_library = libnrnmpi_prefix + "libnrnmpi_" + mpi_implementation + ".so";
-        corenrn_mpi_library = libnrnmpi_prefix + "libcorenrnmpi_" + mpi_implementation + ".so";
-        if (!load_nrnmpi(nrn_mpi_library.c_str(), pmes + strlen(pmes))) {
-            return pmes;
+#endif
+    }();
+
+    // Figure out where to find lib[core]nrnmpi{...} libraries. Older versions
+    // of this code used @loader_path on macOS, which will cause problems in
+    // https://github.com/BlueBrain/CoreNeuron/pull/795 where the code that
+    // calls dlopen(libcorenrnmpi_...) will be in libcorenrnmech.so (in some
+    // model-specific directory) rather than the CoreNEURON installation
+    // directory where libcorenrnmpi_*.so live. Now libcorenrnmech_*.so will be
+    // looked for in the same directory as libnrniv.so, which will be incorrect
+    // if CoreNEURON is built externally with dynamic MPI enabled.
+    auto const libnrnmpi_prefix = []() -> std::string {
+#ifdef MINGW
+        // Preserve old behaviour on Windows
+        return {};
+#else
+        if (const char* nrn_home = std::getenv("NRNHOME")) {
+            // TODO: what about windows path separators?
+            return std::string{nrn_home} + "/lib/";
+        } else {
+            // Use the directory libnrniv.so is in
+            return path_prefix_to_libnrniv();
         }
-    } else {
-        ismes = 1;
-        sprintf(pmes + strlen(pmes),
-                "Is openmpi, mpich, intel-mpi, sgi-mpt etc. installed? If not in default location, "
-                "need a LD_LIBRARY_PATH or MPI_LIB_NRN_PATH.\n");
-    }
-#endif /*not MINGW*/
-#endif /* not DARWIN */
-    if (!handle) {
-        sprintf(pmes + strlen(pmes), "could not dynamically load libmpi.so or libmpich.so\n");
+#endif
+    }();
+
+
+    auto const mpi_path = [&](std::string_view middle) {
+        std::string name{libnrnmpi_prefix};
+        name.append(neuron::config::shared_library_prefix);
+        name.append(middle);
+        name.append(mpi_implementation);
+        name.append(neuron::config::shared_library_suffix);
+        return name;
+    };
+    auto const nrn_mpi_library = mpi_path("nrnmpi_");
+    // TODO this will be wrong if CoreNEURON is installed externally
```

Comment:
Flagging this too, but I'm not sure that CoreNEURON + dynamic MPI in a standalone installation is an important use-case; CoreNEURON is particularly dependent on NEURON when dynamic MPI is enabled.